grpcbp/server_middlewares.go

```diff
@@ -0,0 +1,94 @@
+package grpcbp
+
+import (
+	"context"
+	"strings"
+
+	"github.com/reddit/baseplate.go/ecinterface"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/reddit/baseplate.go/transport"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+func InjectServerSpanInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		split := strings.Split(info.FullMethod, "/")
+		m := split[len(split)-1]
+		ctx, span := StartSpanFromGRPCContext(ctx, m)
+
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			if values := md.Get(transport.HeaderTracingTrace); len(values) > 0 {
+				span.SetTag(tracing.TagKeyPeerService, values[0])
+			}
+		}
+
+		defer func() {
+			span.FinishWithOptions(tracing.FinishOptions{
+				Ctx: ctx,
+				Err: err,
+			}.Convert())
+		}()
+		return handler(ctx, req)
+	}
+}
+
+func InjectEdgeContextInterceptor(impl ecinterface.Interface) grpc.UnaryServerInterceptor {
+	if impl == nil {
+		impl = ecinterface.Get()
+	}
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		ctx = InitializeEdgeContext(ctx, impl)
+		return handler(ctx, req)
+	}
+}
+
+func InitializeEdgeContext(ctx context.Context, impl ecinterface.Interface) context.Context {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ctx
+	}
+	values := md.Get(transport.HeaderEdgeRequest)
+	if len(values) == 0 {
+		return ctx
+	}
+
+	ctx, err := impl.HeaderToContext(ctx, values[0])
+	if err != nil {
+		log.C(ctx).Errorw("Error while parsing EdgeRequestContext: " + err.Error())
```

Comment:
this should be fixed in [`thriftbp`](https://github.com/reddit/baseplate.go/blob/21c209724c85ade8443509934da6543d5ad6744e/thriftbp/server_middlewares.go#L174) as well, but since we are using log directly (not using `log.Wrapper`), it should be:

```suggestion
		log.C(ctx).Errorw("Error while parsing EdgeRequestContext", "err", err)
```

instead.